src/main/Main.hs

```diff
@@ -470,21 +473,41 @@ setupCmd SetupCmdOpts{..} go@GlobalOpts{..} = do
               )
           Nothing
 
-withConfig :: GlobalOpts
+-- | Enforce mutual exclusion of every action running via this function
+-- on this users account.  Currently, stack uses this to completely
+-- exclude concurrent execution of stack commands.  In the future,
+-- stack may refine this to a finer-grained locking approach.
+withUserFileLock :: Config
+                 -> (FileLock -> IO a)
+                 -> IO a
+withUserFileLock cfg act = do
+    lockfile <- parseRelFile "lockfile"
+    let pth = configStackRoot cfg </> lockfile
+    fstTry <- tryLockFile (toFilePath pth) Exclusive
```

Comment:
Technically this code isn't async-exception safe. I'm not sure how important that safety is to the algorithm.